Upload widget component

agbanusi opened the pull request:
## Description
I created the Upload Widget based on the examples given at 
- https://github.com/colbyfayock/cloudinary-examples/tree/main/examples/nextjs-upload-widget-signed
- https://github.com/colbyfayock/cloudinary-examples/tree/main/examples/nextjs-upload-widget-preset

I also added an Upload Button to further abstract the Upload Widget for ease of use for users.

This is the first draft @colbyfayock, please could you help me take a look to ensure I'm in the right direction?
Also apologies for my  formatting from my prettier that changed some js and CSS files

Bugs left to fix:
- [x] only one widget gets triggered, if multiple widgets are rendered on one page
- [x] better error update on signed widgets
- [ ] revert formatting

---
## Issue Ticket Number
Fixes [Issue #4 ]

---
## Type of change
<!-- Please select all options that are applicable. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

---
# Checklist:
- [x] I have followed the contributing guidelines of this project as mentioned in [CONTRIBUTING.md](/CONTRIBUTING.md)
- [x] I have created an [issue](https://github.com/colbyfayock/next-cloudinary/issues) ticket for this PR
- [x] I have checked to ensure there aren't other open [Pull Requests](https://github.com/colbyfayock/next-cloudinary/pulls) for the same update/change?
- [x] I have performed a self-review of my own code
- [ ] I have run tests locally to ensure they all pass
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes needed to the documentation

colbyfayock commented on example/pages/index.js:
```diff
@@ -97,69 +95,76 @@ export default function Home() {
-          <pre><code>{`NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME="<Your Cloud Name>"`}</code></pre>
+          <pre>
+            <code>{`NEXT_PUBLIC_CLOUDINARY_CLOUD_NAME="<Your Cloud Name>"`}</code>
+            <code>{`NEXT_PUBLIC_CLOUDINARY_UPLOAD_PRESET="<Your Preset>"`}</code>
```
i think it would make sense to include the additional environment variables along with the Upload Widget documentation instead of including it here, since it adds an extra unnecessary step if someone only wants to use the image

agbanusi replied:
Yes will remove this and just add to documentation instead

colbyfayock commented on next-cloudinary/src/components/UploadWidget/UploadButton.js:
```diff
@@ -0,0 +1,23 @@
+          }
+          return (
+            <button onClick={handleOnClick} {...props}>
+              {label || "Upload an Image"}
```
what do you think about using `children` for this instead of a `label` prop? my thinking is it gives people more flexibility for how they customize the button

colbyfayock commented on next-cloudinary/src/components/UploadWidget/UploadWidget.js:
```diff
@@ -0,0 +1,103 @@
+   * @description Makes a request to an endpoint to sign Cloudinary parameters as part of widget creation
+   */
+
+  function generateSignature(callback, paramsToSign) {
```
i'm thinking we're going to need to abstract this and actually avoid using a hard-coded endpoint, since it could vary depending on how someone wants to handle this

maybe we create a prop called `signingEndpoint` or something along those lines, and pass in the URL that way? that way they're not forced to create that specific URL. and document what is posted and what's expected to be returned

colbyfayock commented on next-cloudinary/src/components/UploadWidget/UploadWidget.js:
```diff
@@ -0,0 +1,103 @@
+      }),
+    };
+
+    if (signed && !totalOptions.apiKey) {
```
should this be:
```
if (signed && ( !totalOptions.apiKey || !totalOptions.apiSecret ) ) {
```
?

agbanusi replied:
apiSecret is not needed as we only input apiKey ad secret is used in server side API instead

colbyfayock commented on next-cloudinary/src/index.js:
```diff
@@ -1,3 +1,7 @@
+export { default as CldImage } from "./components/CldImage";
+export {
+  default as UploadWidget,
+  UploadButton,
```
this need sto be `default as UploadButton` right?

agbanusi replied:
Only one default is allowed per module in index.js so I did the default as UploadButton there instead

colbyfayock commented:
thanks for this first draft
* As you noted, lots of formatting changes from your internal formatter, can you revert these please?
* love the UploadButton abstraction
* Unsigned Upload Widget worked

getting this error clicking Signed button, though is this what you meant about multiple widgets?
<img width="1177" alt="image" src="https://user-images.githubusercontent.com/1045274/195617926-7817f880-6c36-405f-969f-975036e2ebd8.png">

the Unsigned button click doesnt work, still may be related to above

looking great so far!! i'll think about this a little more as you continue to work on it

agbanusi commented:
> thanks for this first draft
> 
> * As you noted, lots of formatting changes from your internal formatter, can you revert these please?
> * love the UploadButton abstraction
> * Unsigned Upload Widget worked
> 
> getting this error clicking Signed button, though is this what you meant about multiple widgets? <img alt="image" width="1177" src="https://user-images.githubusercontent.com/1045274/195617926-7817f880-6c36-405f-969f-975036e2ebd8.png">
> 
> the Unsigned button click doesnt work, still may be related to above
> 
> looking great so far!! i'll think about this a little more as you continue to work on it

Okay, will revert the format changes and fix the signed button error and the bugs listed above

colbyfayock commented:
hey @agbanusi any luck here?

agbanusi commented:
Apologies for the late response been caught up with a lot. I have fixed all issues except the index.js file formatting which I'd work on shortly.

List of things done are
- I used random values between 0-99 for an attachment to the id and fix the issue of buttons not coming up if multiple widgets or buttons are used on the same page
- I added children to the uploadButton and signatureEndpoint
- added signatureEndpoint to widget instead of hardcoded value
- fixed signed button error of ` widget.current.open not a function`
- tested all buttons and other minor changes

agbanusi commented:
It seems you deleted some files and that is causing conflict on the PR. Should I go ahead and delete them on my PR or do you want to test and confirm then I delete the files?

colbyfayock commented:
awesome ill look at this shortly, but to answer your question, dont worry about it, ill handle the conflicts. ive been making a lot of changes so dont want to put that on you, but thanks

colbyfayock commented:
thanks fort he clarification on some of those points, my mind was elsewhere on a 1 or 2 of those 👀 

this is great though. just a heads up im gonna likely shift things around a little bit. i have a new docs site that 'll update this all on to and just a few organizational preferences i have that i'll make changes for

nice work with all of this!

i'll be merging this in later once i have an opportunity to make the changes but in the meantime

email me at hello@colbyfayock.com where i'll send you a form to claim your free swag! https://spacejelly.dev/hacktoberfest/

colbyfayock commented:
I gotta work more on the documentation but I got it all pulled into the new docs pages! 🥳 

i still have to update some of the pages to add configuration details and such but happy with where it's at to get it merged in

agbanusi commented:
> I gotta work more on the documentation but I got it all pulled into the new docs pages! 🥳
> 
> i still have to update some of the pages to add configuration details and such but happy with where it's at to get it merged in

Thank you very much for the approval and your help 😊
